endaq/device/base.py

```diff
@@ -98,33 +92,66 @@ class Recorder:
     _NAME_PATTERN = re.compile(r'')
 
 
-    def __init__(self, path: Optional[Filename], strict=True):
-        """ Constructor. Typically, instantiation should be done indirectly,
-            using functions such as `endaq.device.getDevices()` or
-            `endaq.device.fromRecording()`. Explicitly instantiating a
-            `Recorder` or `Recorder` subclass is rarely (if ever) necessary.
+    def __init__(self,
+                 path: Optional[Filename],
+                 strict: bool = True,
+                 devinfo: Optional[bytes] = None,
+                 virtual: bool = False):
+        """ A representation of an enDAQ/SlamStick data recorder. Typically,
+            instantiation should be done indirectly, using functions such as
+            `endaq.device.getDevices()` or `endaq.device.fromRecording()`.
+            Explicitly instantiating a `Recorder` or `Recorder` subclass is
+            rarely (if ever) necessary.
 
             :param path: The filesystem path to the recorder, or `None` if
-                it is a 'virtual' device (e.g., constructed from data in a
-                recording).
+                it is a 'virtual' or remote device.
             :param strict: If `True`, only allow real device paths. If
-                `False`, allow any path that contains the standard contents
-                of a recorder's ``SYSTEM`` directory. Primarily for testing.
+                `False`, allow any path that contains a ``SYSTEM`` directory
+                with the standard contents on a device. Primarily for
+                testing.
+            :param devinfo: The necessary data to instantiate a `Recorder`.
+                For creating `Recorder` instances when the hardware isn't
+                physically present on the host computer. If `None`, the
+                data will be read from the device.
+            :param virtual: `True` if the device is not actual hardware
+                (e.g., constructed from data in a recording).
         """
-        # self.mideSchema = loadSchema("mide_ide.xml")
-        # self.manifestSchema = loadSchema("mide_manifest.xml")
-
         self._busy = RLock()
-        self.strict = strict
+        self.strict: bool = strict
+
+        self._virtual: bool = virtual
+        self._command: Optional[CommandInterface] = None
+        self._config: Optional[ConfigInterface] = None
+        self._path: Optional[Filename] = None
+        self._devinfo: Optional[DeviceInfo] = None
+        self._rawinfo: Optional[bytes] = devinfo
+        self._info: Optional[Dict] = None
 
-        self._virtual = False
-        self._command = None
-        self._path = None
-        self.refresh(force=True)
+        self.refresh(force=False)
         self.path = path
+        self.getInfo()
 
         # The source IDE `Dataset` used for 'virtual' devices.
-        self._source = None
+        self._source: Optional[Dataset] = None
+
+
+    def _getDevinfo(self) -> Optional[DeviceInfo]:
```

Comment:
Minor note, it looks like we always return a devinfo and raise an error if it would be a None